### ISSUE TITLE
Simpler examples for the community formatting guidelines

### ISSUE THREAD

**Issue by bbatsov** [kkinnear/zprint#165]
Yesterday I added this section to the community style guide https://guide.clojure.style/#tools and I noticed this page in `zprint`'s docs https://github.com/kkinnear/zprint/blob/master/doc/options/community.md

Different perspective aside (as to what's common sense/consistent and what's a special case), I did struggle to understand the examples that were given, mostly because they were bigger and more complex than they needed to be to illustrate the differences. I also have to mention, that while some people employ wide `cond` formatting with some indentation for the clauses, I think it's way more common to use narrow formatting like this:

```clj
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil
    (not (even? (count clauses)))
    (throw (ex-info (str `cond-let " requires an even number of forms")
                    {:form &form, :meta (meta &form)}))
    :else
    (let [[test expr-or-binding-form & more-clauses] clauses]
      (if (= :let test)
        `(let ~expr-or-binding-form (cond-let ~@more-clauses))
        ;; Standard case
        `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))

;; we can also add a bit of whitespace for clarity
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil

    (not (even? (count clauses)))
    (throw (ex-info (str `cond-let " requires an even number of forms")
                    {:form &form, :meta (meta &form)}))

    :else
    (let [[test expr-or-binding-form & more-clauses] clauses]
      (if (= :let test)
        `(let ~expr-or-binding-form (cond-let ~@more-clauses))
        ;; Standard case
        `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))
```

At least in my experience I've encountered the narrow formatting more often and it seems to me it's easier to parse for humans (as we read better vertically).

**Comment by kkinnear** [comment]
Thanks for the reference to zprint and the community guidelines in the style guide. I'll see what I can do to make the community examples clearer by including some more trivial (but perhaps clearer) examples.  In particular I agree the second example, about body functions and arg functions, is pretty obscure.

Regarding formatting `cond` functions in a wider or narrower format -- zprint was developed in part to try to reduce the cognitive load while editing Clojure source.  In my view, that means being able to see more code on the screen at the same time.   Certainly all editors and IDE's will allow you to zip around and look at things, but being able to see a contiguous chunk of code all at once certainly reduces the cognitive load for me and I believe that I'm not the only person for which this is true. Thus, zprint will format things to reduce the vertical space taken up by the code, while still looking as good as hand formatted code.  Often the results of zprint's formatting are identical to hand formatted code.

I haven't found vertically formatted `cond` functions to be more common than more horizontally formatted `cond` functions, but perhaps I just see the vertically formatted ones as a mistake that someone didn't take the time to correct, as opposed to something done intentionally.  Personally, I find that vertically formatted `cond` functions to be hard to read and rather jarring.  I guess I don't read better vertically.

That said, zprint is designed to allow you to format code the way you want, not restrict you to format it the way that I want.

So, zprint will  easily format things like your first example:
```clojure
; cl is a string that contains the source of the cond-let macro above

(czprint cl {:parse-string? true :style :community :pair {:hang? false}})
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil
    (not (even? (count clauses)))
    (throw (ex-info (str `cond-let " requires an even number of forms")
                    {:form &form, :meta (meta &form)}))
    :else
    (let [[test expr-or-binding-form & more-clauses] clauses]
      (if (= :let test)
        `(let ~expr-or-binding-form (cond-let ~@more-clauses))
        ;; Standard case
        `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))
```
Today zprint will come close to your second example, but will not quite match it.  It will only place blank lines after pairs where the second element of the pair didn't fit on the same line as the first element of the pair, which only occurs once in this example:
```clojure
(czprint cl {:parse-string? true :style :community :pair {:hang? false :nl-separator? true}})
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil
    (not (even? (count clauses)))
    (throw (ex-info (str `cond-let " requires an even number of forms")
                    {:form &form, :meta (meta &form)}))

    :else
    (let [[test expr-or-binding-form & more-clauses] clauses]
      (if (= :let test)
        `(let ~expr-or-binding-form (cond-let ~@more-clauses))
        ;; Standard case
        `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))
```
I will be adding a new feature, `:nl-separator-all?`, which will exactly reproduce your second example, but this won't work until the next release.  I have seen people using this approach, and it is time that zprint supported it exactly:
```clojure
(czprint cl {:parse-string? true :style :community :pair {:hang? false :nl-separator-all? true}})
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil

    (not (even? (count clauses)))
    (throw (ex-info (str `cond-let " requires an even number of forms")
                    {:form &form, :meta (meta &form)}))

    :else
    (let [[test expr-or-binding-form & more-clauses] clauses]
      (if (= :let test)
        `(let ~expr-or-binding-form (cond-let ~@more-clauses))
        ;; Standard case
        `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))
```
Another approach to making the narrow version clearer is to allow indentation for the second element of the pair:
```clojure
(czprint cl {:parse-string? true :pair {:hang? false}})
(defmacro cond-let
  "An alternative to `clojure.core/cond` where instead of a test/expression pair, it is possible
  to have a :let/binding vector pair."
  [& clauses]
  (cond
    (empty? clauses) nil
    (not (even? (count clauses)))
      (throw (ex-info (str `cond-let " requires an even number of forms")
                      {:form &form, :meta (meta &form)}))
    :else
      (let [[test expr-or-binding-form & more-clauses] clauses]
        (if (= :let test)
          `(let ~expr-or-binding-form (cond-let ~@more-clauses))
          ;; Standard case
          `(if ~test ~expr-or-binding-form (cond-let ~@more-clauses))))))
```
but this isn't part of the community guidelines.

I will consider adding a style `:narrow-cond`, and maybe `:narrow-cond-nl` to make it easy for people to produce output that matches these two approaches.

Anyway, thanks very much taking the time to engage with me on these issues!  I appreciate the effort it takes to discuss these things.

**Comment by bbatsov** [comment]
> Thus, zprint will format things to reduce the vertical space taken up by the code, while still looking as good as hand formatted code. Often the results of zprint's formatting are identical to hand formatted code.

Cool. I wasn't aware of that. I must admit that for quite a while I thought of `zprint` only as a `pprint` alternative. 

> That said, zprint is designed to allow you to format code the way you want, not restrict you to format it the way that I want.

I can certainly relate to that approach and I think that's the right approach for the Clojure community. I also like it that you bundled some popular "presets", so people don't have to configure those manually.

> Another approach to making the narrow version clearer is to allow indentation for the second element of the pair:

That's one very interesting approach, that I hadn't encountered so far. I like that it carries more information (which makes it "sematic indentaiton" in my book, but it will requires support in editor formatting engines as well to be viable, as it'd be annoying if the editor changes the formatting you got from zprint. Maybe some editors already support it? 

I'm assuming you're doing doing some similar indentation-wise with everything that takes a binding vector, right? 

 > I haven't found vertically formatted cond functions to be more common than more horizontally formatted cond functions, but perhaps I just see the vertically formatted ones as a mistake that someone didn't take the time to correct, as opposed to something done intentionally. Personally, I find that vertically formatted cond functions to be hard to read and rather jarring. I guess I don't read better vertically.

Yeah, that's certainly not a big deal. Just a random observation of mine, that might be completely inaccurate. I also do a lot of Elisp and there nobody uses wide conds for some reason.

**Comment by kkinnear** [comment]
> I'm assuming you're doing doing some similar indentation-wise with everything that takes a binding vector, right?

Yes, exactly.  It is separately controllable for binding vectors, pairs (`cond`-like things), and maps as well.  

As far as editors go, I'm not clear on that.  I format my source while editing by running portions of my source through the released zprint binaries, so I don't depend on an editor to make it look any particular way.  (The graalVM binaries start in 40-50ms on an 8 year old Macbook Air, so they are plenty fast for this purpose.)  All of which is to say, I don't know of any editors that support this particular approach to indentation.  There are a lot of editors with a lot of plugins and extensions, and it is hard to keep up with even a few of them.  I haven't been able to track them closely for a good while.

**Comment by kkinnear** [comment]
I added better community examples as you suggested in the 1.1.0 release just now.  I also implemented a capability to always place a blank line after every "pair" -- for bindings, `cond` pairs, and maps.  Optionally, of course.  It was an extension to a capability already present to place a blank line after any pair where the second element of the pair ended up on the next line, so it was a straightforward extension.

Thanks for the comments and suggestions!
